pkg/identity/nonglobal/temp_id.go

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package nonglobal
+
+import (
+	"time"
+
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/identity/basicallocator"
+	"github.com/cilium/cilium/pkg/identity/key"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	TempIDGCInterval = 2 * time.Minute
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "non-global-identity")
+)
+
+// TempSecIDAllocator is responsible for allocating temporary security
+// identities. Temporary security identities are used by new pods before a
+// global security identity is allocated by cilium-operator.
+type TempSecIDAllocator struct {
+	allocator          *basicallocator.BasicIDAllocator
+	tempIDCache        *TempIDCache
+	endpointListerFunc func() []*endpoint.Endpoint
+}
+
+// NewTempSecIDAllocator creates a new TempSecIDAllocator for ID values
+// between the provided minIDValue and maxIDValue.
+func NewTempSecIDAllocator(endpointListerFunc func() []*endpoint.Endpoint) *TempSecIDAllocator {
+	return &TempSecIDAllocator{
+		allocator:          basicallocator.NewBasicIDAllocator(identity.DefaultMinTempID, identity.DefaultMaxTempID),
+		tempIDCache:        NewTempIDCache(),
+		endpointListerFunc: endpointListerFunc,
+	}
+}
+
+type TempIDCache struct {
+	mu                lock.RWMutex
+	idToIdentity      map[identity.NumericIdentity]*identity.Identity
+	keyToIdentity     map[string]*identity.Identity
+	markedForDeletion map[identity.NumericIdentity]bool
+}
+
+func NewTempIDCache() *TempIDCache {
+	return &TempIDCache{
+		mu:                lock.RWMutex{},
+		idToIdentity:      make(map[identity.NumericIdentity]*identity.Identity),
+		keyToIdentity:     make(map[string]*identity.Identity),
+		markedForDeletion: make(map[identity.NumericIdentity]bool),
+	}
+}
+
+// FindOrCreateTempID gets ans existing or creates a temporary security identity
+// for the specified labels, to be used before operator creates a global
+// security identity.
+func (a *TempSecIDAllocator) FindOrCreateTempID(lbls labels.Labels) (*identity.Identity, error) {
+	if id, exists := a.LookupByIDKey(&key.GlobalIdentity{LabelArray: lbls.LabelArray()}); exists {
+		return id, nil
+	}
+
+	allocatedID, err := a.allocator.AllocateRandom()
```

Comment:
Is this call thread safe? I guess AllocateRandom is not atomic so what would happen if there is 1 id left and two threads at the same time tries to allocate it?

Reply:
Yes, it's thread safe, because AllocateRandom is, like all methods of BasicIDAllocator, thread safe.

Reply:
I anyway needed to add tempIDCache lock here (moved from insertToCache), to ensure that allocation and cache updates are atomic and don't break GC.